### ISSUE TITLE
The invariant loss does not include a variance term.

### ISSUE THREAD

**Issue by fengyx8** [myZeratul/Causal-Debias#3]
Thank you for your outstanding work.
The current code does not seem to include a variance term when calculating the invariant loss, as it only uses `avg_w_dist`(which refers to merely the expectation term). I'm not sure if there's a problem with my understanding. How can the variance term work in the `avg_w_dist` loss?

The code snippet related to the invariant loss is calculated in the `causal-debias.py` file, lines 777-802.

```
with torch.no_grad():
    if args.model_type == 'roberta':
        output_invariant = model(input_ids, invariant_mask)
        logits_invariant = output_invariant["logits"]
    else:
        output_invariant = model(input_ids, segment_ids, invariant_mask)
        logits_invariant = output_invariant["logits"]


w_dist = 0

for i in range(logits_invariant.shape[0]):
    w_dist += wasserstein_distance(logits_invariant[i].detach().cpu(),logits_ori[i].detach().cpu())
avg_w_dist = w_dist/(logits_ori.shape[0])

loss_fct = CrossEntropyLoss()


loss = loss_fct(logits_ori.view(-1, num_labels), label_ids.view(-1)) + args.tau * avg_w_dist
```

The related equations (5) and (8) demonstrated in `2023 ACL Causal-Debias` paper are as follows:

>
> ![1](https://github.com/myZeratul/Causal-Debias/assets/65220388/104f58b6-ba7d-47af-a754-657ccce5a6eb)
> ![2](https://github.com/myZeratul/Causal-Debias/assets/65220388/c1f638f5-9d9a-4c9e-a8d6-8c8036aba4ce)
>

Looking forward to your reply!

**Comment by myZeratul** [comment]
Sry for updating the wrong version, have a try.
```
w_dist_list = []

for i in range(logits_invariant.shape[0]):
    w_dist_i = wasserstein_distance(
        logits_invariant[i].detach().cpu(),
        logits_ori[i].detach().cpu()
    )
    w_dist_list.append(w_dist_i)

avg_w_dist = torch.mean(torch.tensor(w_dist_list))
var_w_dist = torch.var(torch.tensor(w_dist_list))

loss_fct = cross_entropy

loss = loss_fct(logits_ori.view(-1, num_labels), label_ids.view(-1)) + args.tau * avg_w_dist + args.lambda_var * var_w_dist
```

**Comment by fengyx8** [comment]
Thanks for your reply! @myZeratul 

May I ask what the hyperparameter **`args.lambda_var`** should be set to?

Would you mind just updating your **code** (as well as uploading the indispensable **datasets**) of this repository directly, to ensure that the provided code is **executable** and **correct**?

**Comment by dyi7** [comment]
Hi, 
Thank you for your outstanding work and sharing of codes. 
Following your suggestions in [#1](https://github.com/myZeratul/Causal-Debias/issues/1#issuecomment-1841009702) and [#2](https://github.com/myZeratul/Causal-Debias/issues/3#issuecomment-1841015431), I successfully complied the provided codes without any error. 
Yet when I run the provided `causal-debias.py`, **the loss seems to be NAN**:

  
![Screenshot 2024-06-13 110954](https://github.com/myZeratul/Causal-Debias/assets/44060579/cbcc2de9-a99d-488b-b4b1-6b1ec87b94d1)


I'm not sure if there's a problem with my implementation. I ran `preprocess.py` to preprocess the external corpora, and derived an `external data` file named as `bert_128.bin` (as in https://github.com/myZeratul/Causal-Debias/issues/1#issuecomment-1842864903). Later I run the provided `causal-debias.py`. 

May I ask **how to solve the issue of  `the loss is NAN`** and **get results presented in the paper**?  Looking forward to your reply. Thanks.
